### PR TITLE
[cpd] Improve AnyTokenizer

### LINKED ISSUE #2760: [cpd] AnyTokenizer doesn't count columns correctly
**Affects PMD Version:** 6.25.0+

**Description:**

With 6.25.0 AnyTokenizer now supports columns (b976d6550dfeaf66f32a7efcca2c27beb8610f15), but the columns are not reset when a new line starts. So the beginColumn/endColumn property is always incrementing...

**Code Sample demonstrating the issue:**

Create two files `file1.txt`and `file2.txt` with the following content:

file1.txt
```
file one:
This is some
duplication.
Some other content.
```

file2.txt
```
file2
This is some
    duplication.
Something different.
```

**Steps to reproduce:**

1.  `run.sh cpd --language any --minimum-tokens 4 --format xml --files file1.txt file2.txt`

The output is this:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<pmd-cpd>
   <duplication lines="2" tokens="4">
      <file column="10" endcolumn="33" endline="3" line="2"
            path="/.../file1.txt"/>
      <file column="6" endcolumn="33" endline="3" line="2"
            path="/.../file2.txt"/>
      <codefragment><![CDATA[This is some
duplication.]]></codefragment>
   </duplication>
</pmd-cpd>
```

The begin column should be for both occurrences "1", the endcolumn in file 1 should be 12 and in file 2 it should be 16.

**Running PMD through:** *CLI*

I guess, we simply need to reset the var `colNumber` at https://github.com/pmd/pmd/blob/3ce68f49779ab5185e85e34cf732c353a1d71125/pmd-core/src/main/java/net/sourceforge/pmd/cpd/AnyTokenizer.java#L39

Maybe we can also leverage the new tokenizer test method ("CpdTextComparisonTest") for the AnyTokenizer.

Relates to #2758

### PR DESCRIPTION
**Pull request by oowekyala**
## Describe the PR

* Improve AnyTokenizer a bit. Now it handles string literals and end-of-line comments.
* This makes it as powerful as AbstractTokenizer, albeit less configurable. But configurability is not the goal, and if a user wants a "real" tokenizer with language-specific features, they're anyway better off with a custom lexer built with a lexer generator
* So we can deprecate AbstractTokenizer and remove it in pmd 7. AnyTokenizer is much simpler.

This updates the users of AbstractTokenizer, ie the fortran, perl and ruby modules. It also adds some tests for the perl module, which had none

The difference with the previous AbstractTokenizer implementation for eg ruby, is that no keyword is ignored. I don't think it's a good idea to exclude keywords from the tokens, they're part of the structure of the document. Those tokenizers anyway excluded only a couple of random keywords.

## Related issues

None afaik

* Also fixes #2760 

## Ready?

- [x] Added unit tests for fixed bug/feature
- [x] Passing all unit tests
- [x] Complete build `./mvnw clean verify` passes (checked automatically by travis)
- [x] Added (in-code) documentation (if needed)

### CODE REVIEW

File: pmd-core/src/main/java/net/sourceforge/pmd/util/StringUtil.java [hunk 1]
```diff
@@ -38,6 +38,114 @@
...
+     *
+     * </pre>
+     *
+     * @param charSeq         Char sequence
+     * @param offsetInclusive Offset in the sequence of the targeted character.
+     *                        May be the length of the sequence.
+     * @return -1 if the offset is not in {@code [0, length]}, otherwise
+     * the line number
+     */
+    public static int lineNumberAt(CharSequence charSeq, int offsetInclusive) {
```
Comment (oowekyala) [inline pmd-core/src/main/java/net/sourceforge/pmd/util/StringUtil.java:69]: Note that those utilities and their tests are already on the 7.0.x branch

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">This changeset introduces 0 new violations, 0 new errors and 0 new configuration errors,<br />
removes 0 violations, 0 errors and 0 configuration errors.<br />
<a href="https://chunk.io/pmd/29671c318e72451aa268fff711315c78/diff/index.html">Full report</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: pmd-core/src/main/java/net/sourceforge/pmd/cpd/AnyTokenizer.java [hunk 2]
```diff
@@ -4,44 +4,98 @@
...
+                int bcol = 1 + matcher.start() - lastLineStart; // + 1 because columns are 1 based
+                int ecol = StringUtil.columnNumberAt(image, image.length()); // this already outputs a 1-based column
+                if (ecol == image.length() + 1) {
+                    ecol = bcol + image.length(); // single-line token
+                } else {
+                    // multiline, need to update the line count
+                    lineNo += StringUtil.lineNumberAt(image, image.length()) - 1;
+                    lastLineStart = matcher.start() + image.length() - ecol + 1;
+                }
+                // note: remove this -1 on the 7.0.x branch, this is here because in PMD 6, end columns are inclusive
```
Comment (adangel) [inline pmd-core/src/main/java/net/sourceforge/pmd/cpd/AnyTokenizer.java:89]: TODO

**Review by adangel — APPROVED** [review]
